### PR TITLE
GH-36367: [C++] Add a zipped range utility

### LINKED ISSUE #36367: [C++] Add a zipped range utility
### Describe the enhancement requested

[C++23 adds](https://en.cppreference.com/w/cpp/ranges/zip_view) an analog of Python's zip builtin, enabling parallel iteration of multiple ranges. This doesn't depend on any language features we can't use and would be quite handy for C++ in arrow

### Component(s)

C++

### PR DESCRIPTION
**Pull request by bkietz**
### Rationale for this change

We write a lot of loops over parallel ranges. This function is a proven pattern improving code clarity in other languages

### What changes are included in this PR?

A zip utility is added which simplifies writing loops over parallel ranges.

```diff
@@ -1118,9 +1118,8 @@ Status GetFieldsFromArray(const RjArray& json_fields, FieldPosition parent_pos,
                           DictionaryMemo* dictionary_memo,
                           std::vector<std::shared_ptr<Field>>* fields) {
   fields->resize(json_fields.Size());
-  for (rj::SizeType i = 0; i < json_fields.Size(); ++i) {
-    RETURN_NOT_OK(GetField(json_fields[i], parent_pos.child(static_cast<int>(i)),
-                           dictionary_memo, &(*fields)[i]));
+  for (auto [json_field, field, i] : Zip(json_fields, *fields, Enumerate<int>)) {
+    RETURN_NOT_OK(GetField(json_field, parent_pos.child(i), dictionary_memo, &field));
   }
   return Status::OK();
 }
```

Some(most) of the loops in json_internal.cc are rewritten to showcase usage.

### Are these changes tested?

Yes, in `range_test.cc`.

### Are there any user-facing changes?

No, this is an internal utility.

* Closes: #36367

### CODE REVIEW

File: cpp/src/arrow/util/range.h [hunk 1]
```diff
@@ -151,5 +151,108 @@ LazyRange<Generator> MakeLazyRange(Generator&& gen, int64_t length) {
...
+};
+
+/// \brief A lazy sequence of integers which starts from 0 and never stops.
+///
+/// This can be used in conjunction with Zip() to emulate python's built-in
+/// enumerate() function:
+///
+/// \code {.cpp}
+/// const std::vector<SomeTable>& tables = ...
+/// for (auto [i, table] : Zip(Enumerate<>, tables)) {
```
Comment (pitrou) [inline cpp/src/arrow/util/range.h:230]: For the record, to emulate Python's `enumerate`, I would expect something like:
```suggestion
/// for (auto [i, table] : Zip(Enumerate<>(), tables)) {
```
or:
```suggestion
/// for (auto [i, table] : Zip(Enumerate<>(0), tables)) {
```

Comment (pitrou) [reply cpp/src/arrow/util/range.h:230]: Not necessary useful, though.

**Comment by pitrou** [comment]
(rebased)

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

File: cpp/src/arrow/util/range.h (hunk 1 above)
Comment (bkietz) [reply cpp/src/arrow/util/range.h:230]: Yes, I opted for the simplest option. It'll be easy to add support for an arbitrary start index later if needed.

Related: one thing that unfortunately can't be done with structured bindings is nested unpacking, like
```
for (auto [ [i, table], name ] : Zip(Enumerate(tables), names)) {}
```
which is why I opted to make it strictly a sibling sequence for Zip

Comment (pitrou) [reply cpp/src/arrow/util/range.h:230]: I see, thanks!

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

**Pull request merged**